rfcs/rfc00XX.md

```diff
@@ -0,0 +1,257 @@
+        sub quux {...}
+        sub whee {...}
+    }
+
+
+Imported functions do not remain in the namespace, removing the need for
+namespace::clean and related modules
+
+Upper-case `:export` attribute arguments are reserved for Perl to avoid
+clashing with user-defined tags. This is because in this author's experience,
+export tags are _usually_ lower-case.
+
+## Backwards Compatibility
+
+For basic usage without an `import` method, there are no
+backwards-compatibility issues. There are, however a few surprises.
+
+* Debugger
+
+When using the debugger, it's common to test the return value of a subroutine:
+
+    auto(-1)  DB<2> v
+    30
+    31:    if (@errors) {
+    32:        warn join "\n" => @errors;
+    33==>    say status();
+    34     }
+    35
+    36:    unless ( $opt_for{description} ) {
+    37:        say "Please enter a description for this $opt_for{type} entry:";
+    38:        chomp( my $description = <STDIN> );
+    39:        $opt_for{description} = $description;
+    DB<2> x status()
+    Undefined subroutine &main::status called at ...
+
+The subroutine is bound a compile-time and removed from the namespace. You can
+step into it, but you can't call it directly from the debugger. This should be
+considered.
+
+* Exporting Spec
+
+We often have the following near the top of our code:
+
+    our @EXPORT    = qw(quux);
+    our @EXPORT_OK = qw(
+        bar
+        baz
+        bay
+    )
+    our %EXPORT_TAGS = qw(
+        'all'     => \@EXPORT_OK,
+        'strings' => qw(bay baz),
+    );
+
+Note that there is a need to duplicate some function names.
+
+Some developers like a single exporting spec at the top of a module. Others
+like viewing a function and seeing directly if it can be exported:
+
+    sub foo :export () {
+        ...
+    }
+
+* The `import()` method
+
+This one is tricky because we have to maintain backwards-compatibility.
+
+If there is no `import()` method defined in the class (inheriting does not
+count?), then any attempt to import a function not designated for export should
+be a fatal error:
+
+    $ perl -MMy::Module=whee -E 1
+    "whee" is not exported by the My::Module module
+    Can't continue after import errors at -e line 0.
+    BEGIN failed--compilation aborted.
+
+For the following, we'll use the following stub example of an `import()`
+method:
+
+    package My::Module {
+        use v5.38;
+
+        sub import ($class, @args) {
+            ...
+        }
+        sub bar  :export            {...} # @EXPORT_OK
+        sub baz  :export(:strings)  {...} # %EXPORT_TAGS
+        sub bay  :export(:strings)  {...} # %EXPORT_TAGS
+        sub quux :export(:ALWAYS)   {...} # @EXPORT
+        sub whee                    {...} # not exported
+    }
+
+If an `import()` method is defined, it's possible that it might have special
+importing functionality provided, but we will _still_ allowing importing of
+functions with `:export` tags. We pass the import arguments unchanged to the
```

Comment:
I'd disagree with this. I don't think there should be a new magic layer before `import`.

Instead, I'd say that the exporting should be implemented in UNIVERSAL::import, and if a module wants to use `:export` and have its own import, it should call UNIVERSAL::import in its import method with the appropriate parameters.

Reply:
I admit that your suggestion is much better than mine. However, I'm concerned and I have questions.

First, all procedural modules _inherit_ from `UNIVERSAL`. This is kind of a weird thing which infects _all_ procedural modules. So what happens if a module doesn't use `:export` and calls the superclass `import`? Are there bugs there if the import list mentions a function name? I don't see this as being a problem _per se_, but relying on inheritance for importing is something I would prefer to avoid.

Comments welcome. See @leonerd's comments about an `IMPORT` phaser below.

Reply:
@haarg I've pushed an updated version. In particular, I'm concerned about this part because I think I've missed something.

---
If an `import()` method is defined in the package, it will need to call `UNIVERSAL::import()` to have the lexical importing handled correctly.

```perl
sub import ($class, @args) {
    my ( $import, $mine ) = separate_args(@args);
    # do stuff with $mine
    UNIVERSAL::import($import->@*);
}
```

For the above, how do we tell `UNIVERSAL::import` which class it's importing into? I suppose we could do `local @_ = $import->@*; goto &UNIVERSAL::import`, but that's finicky, it's going to break, and we're trying to discourage use of
`@_` in signatured subs. I expect there's some new work done in Perl which I'm missing here.